transport/src/main/java/io/netty/channel/AbstractChannel.java

```diff
@@ -1053,4 +1079,53 @@ boolean setClosed() {
             return super.trySuccess();
         }
     }
+
+    private static final class AutoFlushTask implements Runnable {
+
+        private final AbstractChannel channel;
+        private final SingleThreadEventLoop eventLoop;
+        private boolean writePending;
+        private boolean running;
+
+        AutoFlushTask(AbstractChannel channel) {
+            this.channel = channel;
+            assert channel.eventLoop() instanceof SingleThreadEventLoop;
+            eventLoop = (SingleThreadEventLoop) channel.eventLoop();
+        }
+
+        @Override
+        public void run() {
+            running = true;
+            try {
+                if (writePending) {
+                    writePending = false;
+                    channel.flush();
+                    if (writePending) {
```

Comment:
It is a bit convoluted, but here is the case (I tried to explain it in the code comment & commit):

If there is a handler which writes more data from within its flush() method, something like:

```java
    @Override
    public void flush(ChannelHandlerContext ctx) throws Exception {
        ctx.flush();
        ctx.write(msg);
    }
```

Then within the call to `channel.flush()`, it will enter into `onWrite()` which will turn the flag back to `true`. This is the case that this piece of code is trying to avoid because it will result in a potential infinite loop of write-flush. By directly flushing the socket (using unsafe) and re-setting the `writePending` flag to `true`, such a loop is broken.